## Nama : Yasmine Alifta
## NIM : 2311110044

# Algoritma Regresi pada Python

jenis machine learning:

1. supervised learning =
2. Unsupervised learning = jenis ML yg tdk mempunyai kelas/tabel pd dataset nya, hanya memiliki variabel x tdk memiliki y
3. reinforcement learning = metode ML yg cara mendapatkan hasilnya berasal dr apa yg dihasilkan dr lingkungan melalui agen

teknik di ML:
1. teknik regularisasi, berfungsi sbg syarat utk menghindari overfitting

- regresi = memprediksi ssuatu nilai bedasarkan garis
- cost function = error tiap value
- loss function = total dari semua error value
- mengukur error bisa dengan = MSE, R^2, MAE
* pada R^2 semaakin mendekati 1 semakin bagus
- uji normalitas
- uji asumsi klasik = mengetes apakah ada masalah linearitas
- uji heterosidasitas
- uji auto korelasi



In [ ]:
from sklearn.datasets import load_diabetes # load dataset diabetes yang sudah disediakan dari sklearn
from sklearn.linear_model import LinearRegression, Lasso, Ridge # mengimpor model regresi untuk variasi regresi regularisasi
from sklearn.model_selection import train_test_split # mengimpor fungsi untuk membagi data menjadi data train dan data test
from sklearn.metrics import mean_squared_error, r2_score # mengimpor metrik untuk mengevaluasi performa model

In [ ]:
x, y = load_diabetes(return_X_y=True) # load data set diabetes ke dalam variable x sebagai feature dan y sebagai target dan akan hanya mengembalikan feature dan target sebagai array

In [ ]:
lr = LinearRegression() # mendefinisikan model regresi linear dan disimpan ke variabel lr

In [ ]:
load_diabetes()['feature_names'] # menampilkan 10 data feature, yaitu 10 variabel yang mempengaruhi target

['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']

In [ ]:
len(load_diabetes()['feature_names']) # menampilkan jumlah total feature yang ada di dataset diabetes, yaitu ada 10 variabel

10

In [ ]:
lr.fit(x, y) # melatih model dengan semua data
y_pred = lr.predict(x) # membuat dan menghasilkan prediksi untuk data x yang disimpan di variabel y_pred

In [ ]:
print(r2_score(y, y_pred)) # menghitung dan menampilkan nilai r2 untuk mengevaluasi seberapa baik model regresi linear dan menjelasskan variabilitas dalam data

0.5177484222203499


hasil R2 score menunjukkan bahwa model dapat menjelaskan sekitar 51.77% variabilitas dalam data target diabetes. Hasil tersebut menujukkan performa yang cukup moderat (tidak terlalu baik, tidak terlalu buruk) dan memiliki ruang untuk peningkatan model.

In [ ]:
print(mean_squared_error(y, y_pred)) # menghitung dan menampilkan nillai MSE untk mrngevaluasi rata-rata error kuadrat antara nilai aktual dan prediksi

2859.69634758675


nilai MSE nya sebesar 2859.69 yang menunjukkan rata-rata kuadrat error prediksi dan nilai aktual. Nilai MSE tersebut cukup besar, sehingga mengindikasikan ada deviasi yang signifikan dalam prediksi.

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state = 42) # membagi dataset menjadi data train dan data test

In [ ]:
lr = LinearRegression() # membuat model regresi linear
lr.fit(x_train, y_train) # melatih model menggunakan data train
y_pred = lr.predict(x_test) # memprediksi nilai target pada data uji
print(r2_score(y_test, y_pred)) # menghitung dan menampilkan hasil r2 score
print(mean_squared_error(y_test, y_pred)) # menghitung dan menampilkan hasil MSE

0.4526027629719195
2900.193628493482


setelah dilakukan split data, nilai R2 score menurun menjadi 45.25% pada data test yang menunjukkan model memiliki kemampuan generalisasi lebih rendah pada data baru.

Sedangkan pada nilai MSE sedikit mengalami peningkatan menjadi 2900.19 yang mengindikasikan error prediksi yang lebih besar pada data test, sehingga model kurang akurat pada data yang belum pernah dilihat.